### PR TITLE
Issue #5624: Google Style Should Enforce Spaces after Commas

### LINKED ISSUE #5624: Google Style Should Enforce Spaces after Commas
`/var/tmp $ javac Test.java`:

```
$ javac Test.java

$
```

`/var/tmp $ cat Test.java`:

```
$ cat Test.java
import java.util.function.BiFunction;

/**
 * A test class to highlight a violation issue.
 */
public class Test {
  /**
   * A lambda, for the purpose of illustrating the issue.
   */
  private static final BiFunction<String,String,String> FUNCTION = (a,b) -> { // Violation -- no space after comma
    return "Hello World, " + a + " and " + b;
  };

  /**
   * Entry point of the application.
   */
  public static void main(String [] args) {
    final Test test = new Test();

    test.doSomething("Lucy","Ricky");
  }

  /**
   * Runs the lambda.
   */
  public void doSomething(final String arg1,final String arg2) { // Violation -- no space after comma
    System.out.println(FUNCTION.apply(arg1,arg2)); // Violation -- no space after comma
  }
}
```

`/var/tmp $ cat config.xml`:

(We're using Google Style at this commit: https://github.com/checkstyle/checkstyle/blob/60f41e3c16e6c94b0bf8c2e5e4b4accf4ad394ab/src/main/resources/google_checks.xml).

`/var/tmp $ java -Duser.language=en -Duser.country=US -jar checkstyle-8.8-all.jar -c config.xml Test.java`:

```
Starting audit...
Audit done.
```

---------------

According to [section 4.6.2 of the official Google Style Guide](https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace), item 5 indicates that "a single ASCII space" should appear after commas, colons, semicolons, "or the closing parenthesis ()) of a cast".

It does not appear that CheckStyle is enforcing this requirement; therefore, there is no warning that all of the commas are missing spaces.

--------------

### PR DESCRIPTION
**Pull request by shashwat70**
Closes #5624: Google Style Should Enforce Spaces after Commas
Diff report: https://shashwat70.github.io/diff2/index.html

### CODE REVIEW

File: src/it/java/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/WhitespaceAfterTest.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Override
+    protected String getPackageLocation() {
+        return "com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace";
+    }
+
+    @Test
+    public void testWhitespaceAfterBasic() throws Exception {
+        final Class<WhitespaceAfterCheck> clazz = WhitespaceAfterCheck.class;
+        final String msgFollowed = "ws.notFollowed";
```
Comment (pbludov) [inline src/it/java/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/WhitespaceAfterTest.java:38]: Please rename `msgFollowed` variable, the name is confusing.

File: src/main/resources/google_checks.xml [hunk 2]
```diff
@@ -48,12 +48,18 @@
...
+            value="COMMA, SEMI, TYPECAST, LITERAL_IF, LITERAL_ELSE,
+            LITERAL_WHILE, LITERAL_DO, LITERAL_FOR, DO_WHILE"/>
+        </module>
         <module name="IllegalTokenText">
             <property name="tokens" value="STRING_LITERAL, CHAR_LITERAL"/>
             <property name="format"
              value="\\u00(09|0(a|A)|0(c|C)|0(d|D)|22|27|5(C|c))|\\(0(10|11|12|14|15|42|47)|134)"/>
             <property name="message"
-             value="Consider using special escape sequence instead of octal value or Unicode escaped value."/>
+             value="Consider using special escape sequence instead of octal value
```
Comment (pbludov) [inline src/main/resources/google_checks.xml:61]: Please revert this change. The violation should be suppressed in https://github.com/checkstyle/checkstyle/blob/aa5a9fd19a6c2e05dbfa756e259f7f59b7d676ce/config/suppressions.xml#L9

File: src/main/resources/google_checks.xml [hunk 3]
```diff
@@ -116,7 +122,8 @@
                      NOT_EQUAL, PLUS, PLUS_ASSIGN, QUESTION, RCURLY, SL, SLIST, SL_ASSIGN, SR,
                      SR_ASSIGN, STAR, STAR_ASSIGN, LITERAL_ASSERT, TYPE_EXTENSION_AND"/>
             <message key="ws.notFollowed"
-             value="WhitespaceAround: ''{0}'' is not followed by whitespace. Empty blocks may only be represented as '{}' when not part of a multi-block statement (4.1.3)"/>
+             value="WhitespaceAround: ''{0}'' is not followed by whitespace. Empty blocks may
```
Comment (pbludov) [inline src/main/resources/google_checks.xml:125]: same here

File: src/it/java/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/WhitespaceAfterTest.java (hunk 1 above)
Comment (shashwat70) [reply src/it/java/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/WhitespaceAfterTest.java:38]: Done

File: src/main/resources/google_checks.xml (hunk 3 above)
Comment (shashwat70) [reply src/main/resources/google_checks.xml:125]: Done

File: src/main/resources/google_checks.xml (hunk 2 above)
Comment (shashwat70) [reply src/main/resources/google_checks.xml:61]: Done

**Comment by rnveach** [comment]
@shashwat70 Regression must be provided for google_checks.xml and the changes being introduced.

File: src/main/resources/google_checks.xml [hunk 4]
```diff
@@ -48,6 +48,11 @@
 
     <module name="TreeWalker">
         <module name="OuterTypeFilename"/>
+        <module name="WhitespaceAfter">
+            <property name="tokens"
+            value="COMMA, SEMI, TYPECAST, LITERAL_IF, LITERAL_ELSE,
+            LITERAL_WHILE, LITERAL_DO, LITERAL_FOR, DO_WHILE"/>
```
Comment (rnveach) [inline src/main/resources/google_checks.xml:54]: How did you identify this list of tokens to add? Issue only talks about commas.

Comment (shashwat70) [reply src/main/resources/google_checks.xml:54]: Yes, it's true that the issue only talks about commas but the actual google style deals with all these elements and requires them to be followed with a horizontal white-space. Please correct me if I'm wrong. Thanks for the review!
https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace

Comment (rnveach) [reply src/main/resources/google_checks.xml:54]: I have not fully looked into the style guide, but if we are doing more than what the issue talks about then the issue needs to be updated to encompass all our changes. In the end, the issue should match what the PR is doing. Issue and issue title is how our users know what we changed as that is what goes on release notes.

Once regression is provided, I will start looking at this in closer details.

Comment (rnveach) [reply src/main/resources/google_checks.xml:54]: Since 4.6.2 is numbered, it will be helpful and speed up review if you posted which tokens applies to which number.

Comment (shashwat70) [reply src/main/resources/google_checks.xml:54]: Alright, I'll try to list them.
1. LITERAL_IF, LITERAL_FOR, LITERAL_WHILE, LITERAL_DO, DO_WHILE
2. LITERAL_ELSE
5. COMMA, SEMI, TYPECAST

**Comment by shashwat70** [comment]
> @shashwat70 Regression must be provided for google_checks.xml and the changes being introduced.

What should be put in the baseConfig and patchConfig properties while generating diff report? @rnveach

File: src/main/resources/google_checks.xml (hunk 4 above)
Comment (rnveach) [reply src/main/resources/google_checks.xml:54]: > 4.6.2.2. LITERAL_ELSE

This seems covered by 4.6.2.3 as 4.6.2.2 is between `}` and `else`.

> TYPECAST

This is the only one I am not sure of because I am not sure of this check, but regression will show what it picks up for it.

I don't see anything else with the tokens added. Other tokens could be added to the check if it supported them.

**Comment by shashwat70** [comment]
I'm facing a problem while generating diff report. I entered the command ``` groovy diff.groovy ``` with all the properties and hit enter, but then nothing happens. I waited for 30 minutes but nothing showed up. It used to work perfectly a week back. What might be the problem? @rnveach

**Comment by shashwat70** [comment]
I'm getting the following error while generating diff report. Please help. @rnveach 
https://gist.github.com/shashwat70/6a3157121a8b3ac5de88e869bdf65adb

**Comment by shashwat70** [comment]
@rnveach It's giving java.lang.OutOfMemoryError also.

**Comment by rnveach** [comment]
@shashwat70 Yea, looks like the problem is related to that.

https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#command-line-arguments
Please use mode `single` and trim down the config to just your additions to `google_checks.xml`.

**Comment by shashwat70** [comment]
Done. The report's link is in the PR description. @rnveach

File: src/main/resources/google_checks.xml [hunk 5]
```diff
@@ -48,6 +48,11 @@
 
     <module name="TreeWalker">
         <module name="OuterTypeFilename"/>
+        <module name="WhitespaceAfter">
+            <property name="tokens"
+            value="COMMA, SEMI, TYPECAST, LITERAL_IF, LITERAL_ELSE,
```
Comment (rnveach) [inline src/main/resources/google_checks.xml:53]: please indent value text with property text.

File: src/main/resources/google_checks.xml (hunk 4 above)
Comment (rnveach) [inline src/main/resources/google_checks.xml:54]: Please indent LITERAL_WHILE to be aligned with COMMA above it.

File: src/xdocs/google_style.xml [hunk 6]
```diff
@@ -970,6 +970,13 @@
                   <span class="wrapper inline">
                     <img src="images/ok_green.png" alt="" />
                   </span>
+                  <a href="config_whitespace.html#WhitespaceAfter">
+                    WhitespaceAfter</a>
+                  <br />
+                  <br />
+                  <span class="wrapper inline">
+                    <img src="images/ok_green.png" alt="" />
```
Comment (rnveach) [inline src/xdocs/google_style.xml:978]: @romani ping

I don't feel this should marked as fully supported. According to https://github.com/checkstyle/checkstyle/pull/7988#discussion_r404379953 other tokens could be added which Google specifies.

File: src/main/resources/google_checks.xml (hunk 4 above)
Comment (rnveach) [reply src/main/resources/google_checks.xml:54]: https://shashwat70.github.io/diff2/guava/index.html
Guava shows no violations with new additions.

File: src/main/resources/google_checks.xml (hunk 5 above)
Comment (shashwat70) [reply src/main/resources/google_checks.xml]: Done

File: src/main/resources/google_checks.xml (hunk 4 above)
Comment (shashwat70) [reply src/main/resources/google_checks.xml]: Done

File: src/xdocs/google_style.xml (hunk 6 above)
Comment (shashwat70) [reply src/xdocs/google_style.xml:978]: Yes, I guess you're right. Should I change it? @rnveach

Comment (romani) [reply src/xdocs/google_style.xml:978]: if we can not 100% reliably validate it , it is better to make it "blue", in issue please make a comment on what cases we can miss or mishandle.

Comment (rnveach) [reply src/xdocs/google_style.xml:978]: yes @shashwat70

Comment (shashwat70) [reply src/xdocs/google_style.xml]: Done. @rnveach

File: src/it/resources/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/InputWhitespaceAfter.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
+package com.google.checkstyle.test.chapter4formatting.rule462horizontalwhitespace;
+
+public class InputWhitespaceAfter {
+    public void check1(int x,int y){ //warn
+        for(int a=1,b=2;a<5;a++,b--); // warn
```
Comment (romani) [inline src/it/resources/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/InputWhitespaceAfter.java:5]: please go through regression report and add more test case, like https://shashwat70.github.io/diff2/Vavr/index.html#A45

There should test case for valid and invalid for each point from https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace
Please create valid Input file with all case and invalid Input file with violations (ideally such Inputs should almost identical and diff only in missed spaces in invalid file).

File: src/main/resources/google_checks.xml [hunk 8]
```diff
@@ -48,6 +48,11 @@
 
     <module name="TreeWalker">
         <module name="OuterTypeFilename"/>
+        <module name="WhitespaceAfter">
```
Comment (romani) [inline src/main/resources/google_checks.xml:51]: please move it to be right after/before `WhitespaceAround` to at least group Checks configs a bit.

File: src/it/resources/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/InputWhitespaceAfter.java (hunk 7 above)
Comment (romani) [reply src/it/resources/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/InputWhitespaceAfter.java:5]: If you do not want to add all tokens and tests for them to completely cover style chapter, as in original issue only comas were reported. It is ok, but please create separate issue on items that we miss even after your fix.

Comment (shashwat70) [reply src/it/resources/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/InputWhitespaceAfter.java:5]: > please create separate issue on items that we miss even after your fix.

moved to issue #8122 
I'll update its issue description in a few moments.

Comment (romani) [reply src/it/resources/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/InputWhitespaceAfter.java:5]: All tokens that you adding in this PR, please cover by valid and invalid test cases

Comment (shashwat70) [reply src/it/resources/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/InputWhitespaceAfter.java]: Done

File: src/main/resources/google_checks.xml (hunk 8 above)
Comment (shashwat70) [reply src/main/resources/google_checks.xml]: Done

File: src/it/resources/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/InputWhitespaceAfter.java (hunk 7 above)
Comment (shashwat70) [reply src/it/resources/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/InputWhitespaceAfter.java]: I've added an example to the new issue. Please approve it and add required tags so that I may try to work on it. @romani

Comment (shashwat70) [reply src/it/resources/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/InputWhitespaceAfter.java]: My pleasure :)
